File: src/javascript/lib/agent.js
import Kernel from './kernel';
import Keyword from './keyword';

let Agent = {};

Agent.start = function(fun, options = []){
  let pid = self.processes.spawn();

  if(Keyword.has_key__qm__(options, Kernel.SpecialForms.atom("name"))){
    pid = self.processes.register(Keyword.get(options, Kernel.SpecialForms.atom("name")), pid)
  }

  self.processes.put(pid, "state", fun());
  return Kernel.SpecialForms.tuple(Kernel.SpecialForms.atom("ok"), pid);
}

Agent.stop = function(agent, timeout = 5000){
  self.processes.exit(agent);
  return Kernel.SpecialForms.atom("ok");
}

Agent.update = function(agent, fun, timeout = 5000){

  const current_state = self.processes.get(agent, "state");
  self.processes.put(agent, "state", fun(current_state));

  return Kernel.SpecialForms.atom("ok");
}

Agent.get = function(agent, fun, timeout = 5000){
  return fun(self.processes.get(agent, "state"));
}

Agent.get_and_update = function(agent, fun, timeout = 5000){

  const get_and_update_tuple = fun(self.processes.get(agent, "state"));
  self.processes.put(agent, "state", Kernel.elem(get_and_update_tuple, 1));

  return Kernel.elem(get_and_update_tuple, 0);
}

export default Agent;

File: lib/elixir_script/prelude/agent.ex
defmodule ElixirScript.Agent do
    @moduledoc false

  def start(fun, options \\ []) do
    pid = spawn(fn() -> Process.sleep(:infinity) end)

    if Elixir.Keyword.has_key?(options, :name) do
      Process.register(pid, Elixir.Keyword.get(options, :name))
    end

    Elixir.Core.Store.create(pid, fun.())
    { :ok, pid }
  end

  def start_link(fun, options \\ []) do
    pid = spawn_link(fn() -> Process.sleep(:infinity) end)

    if Elixir.Keyword.has_key?(options, :name) do
      Process.register(pid, Elixir.Keyword.get(options, :name))
    end

    Elixir.Core.Store.create(pid, fun.())
    { :ok, pid }
  end

  def stop(agent) do
    Process.exit(agent)
    :ok
  end

  def update(agent, fun) do
    pid = Elixir.Core.processes.pidof(agent)
    current_state = Elixir.Core.Store.read(pid)
    Elixir.Core.Store.update(pid, fun.(current_state))
    :ok
  end

  def get(agent, fun) do
    pid = Elixir.Core.processes.pidof(agent)
    current_state = Elixir.Core.Store.read(pid)
    fun.(current_state)
  end

  def get_and_update(agent, fun) do
    pid = Elixir.Core.processes.pidof(agent)
    current_state = Elixir.Core.Store.read(pid)
    {val, new_state} = fun.(current_state)
    Elixir.Core.Store.update(pid, new_state)
    val
  end

end
